### PR TITLE
Travis: report coverage with all builds

### PR DESCRIPTION
**Pull request by blueyed**
https://codecov.io/gh/pytest-dev/pytest/tree/coverage

TODO:

- [x] https://github.com/pytest-dev/pytest/pull/3877 should be merged first, then this one rebased.
- [x] bring back AppVeyor

### CODE REVIEW

**Comment by blueyed** [comment]
It seems to be problematic with py27/py26 (https://travis-ci.org/pytest-dev/pytest/builds/280216101).
I've triggered a full build now again.

Is this something we want to have in general?

File: .travis.yml [hunk 1]
```diff
@@ -5,11 +5,15 @@ python:
 # command to install dependencies
 install:
   - pip install --upgrade --pre tox
+  - |
+    pip install tox
```
Comment (nicoddemus) [inline .travis.yml]: tox is already installed in the line above, so this is not really needed

File: .travis.yml [hunk 2]
```diff
@@ -5,11 +5,15 @@ python:
...
+    if [[ "${TOXENV#py}" != "$TOXENV" ]]; then
```
Comment (nicoddemus) [inline .travis.yml]: Pardon my lack of bash-fu, what does this expression inside `if` does?

File: tox.ini [hunk 3]
```diff
@@ -64,57 +67,65 @@ deps =
...
 [testenv:py27-pexpect]
 changedir = testing
 platform = linux|darwin
-deps = pexpect
+deps =
+    pexpect
+    {env:TOX_EXTRA_DEPS:}
 commands =
-    pytest -ra test_pdb.py test_terminal.py test_unittest.py
+    {env:TOX_COVERAGE_RUN:} pytest -ra test_pdb.py test_terminal.py test_unittest.py
```
Comment (nicoddemus) [inline tox.ini]: Looking at the Travis build, `py27/py36-trial` seem to be missing this configuration

File: .coveragerc [hunk 4]
```diff
@@ -1,4 +1,5 @@
 [run]
+source = _pytest
 omit = 
     # standlonetemplate is read dynamically and tested by test_genscript
```
Comment (nicoddemus) [inline .coveragerc]: Oh this does not exist anymore and can be removed. 😁

File: .travis.yml [hunk 5]
```diff
@@ -5,11 +5,15 @@ python:
...
+      export TOX_COVERAGE_RUN="coverage run -m"
```
Comment (nicoddemus) [inline .travis.yml]: At first glance I thought `TOX_EXTRA_DEPS` was a feature of tox itself. Perhaps we should add a PYTEST_ prefix to avoid confusion? While at it, perhaps there is a way to add extra dependencies using some tox feature?

Comment (nicoddemus) [reply .travis.yml]: Just remembered that we used `_PYTEST_SETUP_SKIP_PLUGGY_DEP` in [setup.py](https://github.com/pytest-dev/pytest/blob/features/setup.py#L49) for this type of "setup/buid only flag used internally".

`_PYTEST_TOX_COVERAGE_RUN`?

File: .coveragerc (hunk 4 above)
Comment (blueyed) [reply .coveragerc]: standlonetemplate?

Apart from that I think we should also include `testing` (the tests themselves).  Anything else?

File: .travis.yml (hunk 2 above)
Comment (blueyed) [reply .travis.yml]: TOXENV starts with "py".

Comment (nicoddemus) [reply .travis.yml]: Thanks, and what does `"${TOXENV#py}"` do? (I tried googling it but it wasn't entirely obvious what to search for).

File: .coveragerc (hunk 4 above)
Comment (nicoddemus) [reply .coveragerc]: "standlonetemplate" was used by the removed `genscript` feature.

> Apart from that I think we should also include testing (the tests themselves). Anything else?

Not sure, then `xfail` tests which are never meant to run because they reproduce issues not yet fixed will drop the coverage. In my head "coverage" does not include the tests. What's your opinion?

**Comment by blueyed** [comment]
Failures on py2, pypy and py33 are all:
```
=========================== short test summary info ============================
FAIL testing/code/test_excinfo.py::TestTraceback_f_g_h::()::test_traceback_cut_excludepath
FAIL testing/python/collect.py::TestModule::()::test_show_traceback_import_error[0]
FAIL testing/python/collect.py::TestModule::()::test_show_traceback_import_error[1]
FAIL testing/python/collect.py::TestTracebackCutting::()::test_traceback_argsetup
```
e.g. https://travis-ci.org/pytest-dev/pytest/jobs/280479131#L599

Any idea?

File: .travis.yml (hunk 5 above)
Comment (blueyed) [reply .travis.yml]: Makes sense.  Will rename it to `_PYTEST_TOX_COVERAGE_RUN` and `_PYTEST_TOX_EXTRA_DEPS` then?

File: .travis.yml (hunk 2 above)
Comment (blueyed) [reply .travis.yml]: It substitutes "py" at the beginning.
Yes, shell programming can be ugly.  See "Parameter Expansion" in bash's man page.

**Comment by blueyed** [comment]
@nicoddemus 
Can you enable the codecov bot user on https://codecov.io/gh/pytest-dev/pytest?

**Comment by nicoddemus** [comment]
> Can you enable the codecov bot user on https://codecov.io/gh/pytest-dev/pytest?

Done! If we will use `codecov` we should disable `coveralls` (after this is merged)

File: .travis.yml (hunk 5 above)
Comment (nicoddemus) [reply .travis.yml]: Sounds good, thanks

File: .travis.yml (hunk 2 above)
Comment (nicoddemus) [reply .travis.yml]: Thanks!

File: .coveragerc (hunk 4 above)
Comment (blueyed) [reply .coveragerc]: I think unexpected missing coverage is a good sign of something being wrong.
Those xfail tests should be marked with "pragma: no cover" then.

**Comment by blueyed** [comment]
The problem with `test_traceback_cut_excludepath` is that the traceback entry looks different when run with `-m`, also when using `python -m pytest` (and we use `coverage run -m`).

`<TracebackEntry _pytest/python_api.py:580>` vs `<TracebackEntry /home/user/Vcs/pytest/.tox/py27/lib/python2.7/site-packages/_pytest/python_api.py:580>`.
Is that something that should get fixed for `excinfo.traceback`?
Looks like it uncovered an actual issue?!

**Comment by blueyed** [comment]
This works around it:

```diff
diff --git i/testing/code/test_excinfo.py w/testing/code/test_excinfo.py
index f8f8a036..cc512a2d 100644
--- i/testing/code/test_excinfo.py
+++ w/testing/code/test_excinfo.py
@@ -162,9 +162,12 @@ def test_traceback_cut(self):
         assert len(newtraceback) == 1
 
     def test_traceback_cut_excludepath(self, testdir):
+        ptf = py.path.local(pytest.__file__)
+        if not ptf.exists():
+            pytest.skip('pytest.__file__ does not point to an actual file.')
+        basedir = ptf.dirpath()
         p = testdir.makepyfile("def f(): raise ValueError")
         excinfo = pytest.raises(ValueError, "p.pyimport().f()")
-        basedir = py.path.local(pytest.__file__).dirpath()
         newtraceback = excinfo.traceback.cut(excludepath=basedir)
         for x in newtraceback:
             if hasattr(x, 'path'):
```

Looks like an issue with Python 2 where `pytest.__file__` is not correct in this case.

**Comment by nicoddemus** [comment]
I gotta say I like codecov's report better. 😄

File: .travis.yml [hunk 6]
```diff
@@ -49,6 +52,18 @@ matrix:
...
 script: tox --recreate
 
+after_success:
+  - |
+    set -ex
+    if [[ "${TOXENV#py}" != "$TOXENV" ]]; then
+      pip install codecov coveralls
+      coverage xml
+      coverage report -m
+      coveralls
```
Comment (nicoddemus) [inline .travis.yml]: Should we drop `coveralls` if we are migrating to `codecov`?

File: tox.ini [hunk 7]
```diff
@@ -19,29 +19,32 @@ envlist =
...
 [testenv:py27-subprocess]
 changedir = .
 deps =
     pytest-xdist>=1.13
     mock
     nose
+    {env:TOX_EXTRA_DEPS:}
 commands =
-    pytest -n3 -ra --runpytest=subprocess {posargs:testing}
+    {env:TOX_COVERAGE_RUN:} pytest -n3 -ra --runpytest=subprocess {posargs:testing}
```
Comment (nicoddemus) [inline tox.ini]: I'm seeing this warning from tox:

```
WARNING:test command found but not installed in testenv
  cmd: /usr/bin/env
  env: /home/travis/build/pytest-dev/pytest/.tox/py27
```

Do we have to whitelist `env`?

File: tox.ini [hunk 8]
```diff
@@ -183,23 +196,13 @@ commands =
...
 deps =
-    {[testenv]deps}
-    coveralls
+    pyinstaller
+    {env:TOX_EXTRA_DEPS:}
 commands =
-    coverage run --source=_pytest -m pytest testing
-    coverage report -m
-    coveralls
+    {env:TOX_COVERAGE_RUN:} {envpython} create_executable.py
```
Comment (nicoddemus) [inline tox.ini]: We can drop coverage from this env I believe

File: tox.ini [hunk 9]
```diff
@@ -19,29 +19,32 @@ envlist =
     docs
 
 [testenv]
-commands = pytest --lsof -ra {posargs:testing}
+commands = {env:TOX_COVERAGE_RUN:} pytest --lsof -ra {posargs:testing}
```
Comment (RonnyPfannschmidt) [inline tox.ini]: can we just introduce a coverage factor for this?

File: .travis.yml [hunk 10]
```diff
@@ -5,11 +5,14 @@ python:
 # command to install dependencies
 install:
   - pip install --upgrade --pre tox
+  - |
+    if [[ "${TOXENV#py}" != "$TOXENV" ]]; then
+      export TOX_COVERAGE_RUN="env COVERAGE_FILE=$PWD/.coverage coverage run -a -m"
```
Comment (RonnyPfannschmidt) [inline .travis.yml]: at frist glance this "breaks" at least for the xdist envs

File: tox.ini [hunk 11]
```diff
@@ -19,29 +19,32 @@ envlist =
...
 passenv = USER USERNAME
 deps =
     hypothesis>=3.5.2
     nose
     mock
     requests
+    {env:TOX_EXTRA_DEPS:}
```
Comment (RonnyPfannschmidt) [inline tox.ini]: dito

File: tox.ini (hunk 7 above)
Comment (The-Compiler) [reply tox.ini]: Couldn't we just use [pytest-cov](https://github.com/pytest-dev/pytest-cov) instead?

Comment (RonnyPfannschmidt) [reply tox.ini]: pytest-cov does incomplete reports due to late plugin startup, its a longstanding pytest issue

File: tox.ini (hunk 9 above)
Comment (blueyed) [reply tox.ini]: @RonnyPfannschmidt 
There is no way to have a command by default and then another for a factor, is there? (where the command comes not from the default testenv).

**Comment by blueyed** [comment]
@nicoddemus 
See https://github.com/pytest-dev/pytest/pull/2800#issuecomment-332635907.

**Comment by RonnyPfannschmidt** [comment]
instead of falling out for pytest in some cases

lets use `_pytest` instead - its not being used as module

**Comment by nicoddemus** [comment]
> lets use _pytest instead - its not being used as module

You mean in the production code or the test?

**Comment by RonnyPfannschmidt** [comment]
in `testing/code/test_excinfo.py`

**Comment by nicoddemus** [comment]
@blueyed I think we have fixed the errors you were experiencing, would you like to give this another shot?

File: .travis.yml (hunk 6 above)
Comment (blueyed) [reply .travis.yml]: I think it is good to keep it for comparison, at least for a while.

File: tox.ini (hunk 9 above)
Comment (blueyed) [reply tox.ini]: Will do it via `setenv`:
```
[testenv]
commands = {env:_TOX_COVERAGE_RUN:} pytest --lsof -ra {posargs:testing}
    coverage: coverage report -m --skip-covered
setenv =
    coverage: _TOX_COVERAGE_RUN=env COVERAGE_FILE={toxinidir}/.coverage coverage run -a -m
whitelist_externals =
    coverage: env
```

**Comment by blueyed** [comment]
@nicoddemus 
The problem(s) with py27 seems to be still there: https://travis-ci.org/pytest-dev/pytest/jobs/314001060#L621.
But I had rebased it against master - now for features.

**Comment by nicoddemus** [comment]
@blueyed gentle ping if you still want to give this a shot; `coveralls` is acting strange this days so `codecov` would be a welcome change. 😁

**Comment by blueyed** [comment]
py27-xdist fails (https://travis-ci.org/pytest-dev/pytest/jobs/420593988#L668):

> Couldn't parse '/home/travis/build/pytest-dev/pytest/.tox/py27-xdist/lib/python2.7/site-packages/nose/config.py' as Python source: 'invalid syntax' at line 68

py36-freeze fails (https://travis-ci.org/pytest-dev/pytest/jobs/420594002#L665):

> No data to report.

We should skip pypy, because it is very slow with coverage reporting (https://travis-ci.org/pytest-dev/pytest/jobs/420594000) - takes > 20mins for 68% already.

File: sitecustomize.py [hunk 12]
```diff
@@ -0,0 +1,9 @@
+"""Setup coverage tracking for subprocesses.
+
+Any ImportError is silently ignored.
+Requires COVERAGE_PROCESS_START in the environments, which gets set in
+tox.ini.
+"""
+import coverage
+
+coverage.process_startup()
```
Comment (asottile) [inline sitecustomize.py]: I've used `coverage-enable-subprocess` with success before -- it essentially does this but with a `.pth` file (no need for a relative `sitecustomize.py` file)

Comment (blueyed) [reply sitecustomize.py]: Awesome, will use that.

File: .travis.yml [hunk 13]
```diff
@@ -62,15 +59,44 @@ jobs:
...
       - pre-commit install-hooks
       script:
       - pre-commit run --all-files
 
+before_script:
+  - |
+    if [[ "$PYTEST_NO_COVERAGE" != 1 ]]; then
+      set -ex
+      # Massage .coveragerc to use source from tox env (only).
+      sed -i "s~source~source = $(ls $PWD/.tox/$TOXENV/lib/python*/site-packages/_pytest/),$PWD/testing" .coveragerc
```
Comment (asottile) [inline .travis.yml]: one trick you can use here is `.coveragerc` supports environment variables.

I'm trying to find a not-closed-source example for you but this is the rough sketch:

```ini
setenv =
    TOP={toxinidir}
```

and then in `.coveragerc`

```
source = $TOP
```

(I think `$PWD` probably works fine too, without `setenv` -- though I'm trying to figure out why different 🤔 )

The rough part here is when things `chdir` you want an absolute path...

**Comment by blueyed** [comment]
Yay! https://codecov.io/gh/pytest-dev/pytest/tree/5abd519e67d7ad889bdbe44fa746af8f3eacac09

Please review the PR via https://github.com/pytest-dev/pytest/pull/2800/files, and I will clean it up after #3877 has been merged.

File: .travis.yml [hunk 14]
```diff
@@ -62,10 +61,35 @@ jobs:
...
+  - |
+    if [[ "$PYTEST_NO_COVERAGE" != 1 ]]; then
+      set -e
+      pip install codecov
+      coverage combine
+      coverage xml
+      coverage report -m
+      codecov --required -X gcov pycov search -f coverage.xml --flags ${TOXENV//-/ }
+
+      # Coveralls does not support merged reports.
```
Comment (nicoddemus) [inline .travis.yml]: I believe we can drop coveralls then. Or would you rather do it in another PR @blueyed?

File: appveyor.yml [hunk 15]
```diff
@@ -1,53 +0,0 @@
-environment:
```
Comment (nicoddemus) [inline appveyor.yml]: I assume you removed AppVeyor so the builds are faster, please remember to bring it back before merging.

File: tox.ini [hunk 16]
```diff
@@ -17,13 +17,21 @@ envlist =
     docs
 
 [testenv]
-commands = pytest --lsof -ra {posargs:testing}
-passenv = USER USERNAME
+commands =
+    {env:_PYTEST_TOX_COVERAGE_RUN:} pytest --lsof -ra {env:_PYTEST_TEST_OPTS:} {posargs:testing}
+    coverage: coverage report -m --skip-covered
+setenv =
+    coverage: _PYTEST_TOX_COVERAGE_RUN=coverage run -a -m
```
Comment (nicoddemus) [inline tox.ini]: Why do we handle `_PYTEST_TOX_COVERAGE_RUN` and  `_PYTEST_TOX_EXTRA_DEP` both here and in the travis file?

Comment (blueyed) [reply tox.ini]: This is meant to allow for `tox -e py36-coverage`.
I've thought about/tried using only `-coverage`, but it is not trivial/feasible due to the many different testenvs.

File: .travis.yml (hunk 14 above)
Comment (blueyed) [reply .travis.yml]: I've thought to keep it for comparison.

Comment (nicoddemus) [reply .travis.yml]: Oh OK. But I think we should remove it after a while to avoid confusion. 👍

File: tox.ini (hunk 16 above)
Comment (nicoddemus) [reply tox.ini]: I see, OK, thanks

**Comment by blueyed** [comment]
Squashed (old HEAD: https://github.com/pytest-dev/pytest/pull/2800/commits/78d2c4deb6e6a44236968faf3c31c88e747b3fc9).
I think this is good from my side for now.

**Comment by blueyed** [comment]
Meh.. AppVeyor will fail - still using TOXENV=coveralls..

I guess the same mechanism should be used with AppVeyor then?  Basically the `before_script` and `after_success` from Travis would need to be ported then.
@nicoddemus can you do it?  It should also add an extra flag with codecov then, "win" or "windows".

**Comment by nicoddemus** [comment]
@blueyed sure I can give it a try. For now I suggest to comment out the `TOXENV=coveralls` in AppVeyor so we can get this merged.

**Comment by blueyed** [comment]
Rebased it on master (where the other ci improvements are already), and re-added coveralls for AppVeyor.  Since it also runs codecov there it should be good for Windows for now.

**Comment by nicoddemus** [comment]
> Since it also runs codecov there 

You mean `coveralls` right?

**Comment by blueyed** [comment]
Both, since 4d19b943.

**Comment by blueyed** [comment]
Fixed linting, squashed the keeping of coveralls for AppVeyor:

```diff
diff --git a/.travis.yml b/.travis.yml
index 9279334e..2c9deb3e 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -60,7 +60,7 @@ jobs:
       env: TOXENV=py27
     - env: TOXENV=py34
     - env: TOXENV=py36
-    - env: TOXENV=linting
+    - env: TOXENV=linting PYTEST_NO_COVERAGE=1

     - stage: deploy
       python: '3.6'
```

**Comment by blueyed** [comment]
No coverage for doctesting, but works locally (https://travis-ci.org/pytest-dev/pytest/jobs/422108111#L625).
Trying to set `COVERAGE_FILE` for it.

**Comment by blueyed** [comment]
doctesting is working now (https://travis-ci.org/pytest-dev/pytest/jobs/422263747#L677), but I've changed the tox testenv quite a bit.

**Comment by nicoddemus** [comment]
Thanks a ton @blueyed!

**Comment by blueyed** [comment]
Yay! :)
